packages/@aws-cdk/aws-route53-patterns/lib/vpc-endpoint-service-domain-name.ts

```diff
@@ -0,0 +1,178 @@
+import { AwsCustomResource, AwsCustomResourcePolicy, PhysicalResourceId } from '@aws-cdk/custom-resources';
+import { Construct } from 'constructs';
+
+// v2 - keep this import as a separate section to reduce merge conflict when forward merging with the v2 branch.
+// eslint-disable-next-line
+import { Construct as CoreConstruct } from '@aws-cdk/core';
+
+/**
+ * Properties to configure a VPC Endpoint Service domain name
+ */
+export interface VpcEndpointServiceDomainNameProps {
+
+  /**
+   * The VPC Endpoint Service to configure Private DNS for
+   */
+  readonly endpointService: IVpcEndpointService;
+
+  /**
+   * The domain name to use.
+   *
+   * This domain name must be verifiably owned by this account, or otherwise
+   * delegated to this account.
+   * https://docs.aws.amazon.com/vpc/latest/userguide/endpoint-services-dns-validation.html
+   */
+  readonly domainName: string;
+
+  /**
+   * The public hosted zone to use for the domain.
+   */
+  readonly publicZone: IPublicHostedZone;
+}
+
+/**
+ * A Private DNS configuration for a VPC endpoint service.
+ */
+export class VpcEndpointServiceDomainName extends CoreConstruct {
+
+  // Track all domain names created, so someone doesn't accidentally associate two domains with a single service
+  private static readonly endpointServices: IVpcEndpointService[] = [];
+
+  // The way this class works is by using three custom resources and a TxtRecord in conjunction
+  // The first custom resource tells the VPC endpoint service to use the given DNS name
+  // The VPC endpoint service will then say:
+  // "ok, create a TXT record using these two values to prove you own the domain"
+  // The second custom resource retrieves these two values from the service
+  // The TxtRecord is created from these two values
+  // The third custom resource tells the VPC Endpoint Service to verify the domain ownership
+  constructor(scope: Construct, id: string, props: VpcEndpointServiceDomainNameProps) {
+    super(scope, id);
+
+    // Make sure a user doesn't accidentally add multiple domains
+    this.validateParams(props);
+    VpcEndpointServiceDomainName.endpointServices.push(props.endpointService);
+
+    const serviceId = props.endpointService.vpcEndpointServiceId;
+    const privateDnsName = props.domainName;
+
+    // Turns creates the Private DNS configuration on the endpoint service
+    const enable = new AwsCustomResource(this, 'EnableDns', {
+      onCreate: {
+        service: 'EC2',
+        action: 'modifyVpcEndpointServiceConfiguration',
+        parameters: {
+          ServiceId: serviceId,
+          PrivateDnsName: privateDnsName,
+        },
+        physicalResourceId: PhysicalResourceId.of(id),
+      },
+      onUpdate: {
+        service: 'EC2',
+        action: 'modifyVpcEndpointServiceConfiguration',
+        parameters: {
+          ServiceId: serviceId,
+          PrivateDnsName: privateDnsName,
+        },
+        physicalResourceId: PhysicalResourceId.of(id),
+      },
```

Comment:
seems like the values used for `onCreate` and `onUpdate` are the same - can we DRY this up a little and refactor?